app/src/main/java/com/duckduckgo/app/location/ui/LocationPermissionsAdapter.kt

```diff
@@ -0,0 +1,290 @@
+import android.widget.ImageView
+import android.widget.PopupMenu
+import androidx.core.view.isGone
+import androidx.recyclerview.widget.RecyclerView
+import com.duckduckgo.app.browser.R
+import com.duckduckgo.app.global.faviconLocation
+import com.duckduckgo.app.global.image.GlideApp
+import com.duckduckgo.app.global.view.quietlySetIsChecked
+import com.duckduckgo.app.global.view.website
+import com.duckduckgo.app.location.data.LocationPermissionEntity
+import com.duckduckgo.app.location.data.LocationPermissionType
+import kotlinx.android.synthetic.main.view_location_permissions_entry.view.locationPermissionEntryDomain
+import kotlinx.android.synthetic.main.view_location_permissions_entry.view.locationPermissionEntryFavicon
+import kotlinx.android.synthetic.main.view_location_permissions_entry.view.locationPermissionMenu
+import kotlinx.android.synthetic.main.view_location_permissions_section_title.view.locationPermissionsSectionTitle
+import kotlinx.android.synthetic.main.view_location_permissions_toggle.view.locationPermissionsToggle
+import timber.log.Timber
+
+class LocationPermissionsAdapter(
+    private val viewModel: LocationPermissionsViewModel
+) : RecyclerView.Adapter<LocationPermissionsViewHolder>() {
+
+    companion object {
+        const val PRECISE_LOCATION_DOMAIN_TYPE = 0
+        const val DESCRIPTION_TYPE = 1
+        const val EMPTY_STATE_TYPE = 2
+        const val TOGGLE_TYPE = 3
+        const val DIVIDER_TYPE = 4
+        const val ALLOWED_SITES_SECTION_TITLE_TYPE = 5
+        const val DENIED_SITES_SECTION_TITLE_TYPE = 6
+
+        const val EMPTY_HINT_ITEM_SIZE = 1
+    }
+
+    fun updatePermissions(
+        isLocationPermissionEnabled: Boolean,
+        newPermissions: List<LocationPermissionEntity>
+    ) {
+        locationPermissionEnabled = isLocationPermissionEnabled
+        locationPermissions = newPermissions
+        notifyDataSetChanged()
+
+        // val diffCallback = DiffCallback(locationPermissions, newPermissions)
+        // val diffResult = DiffUtil.calculateDiff(diffCallback)
+        // locationPermissions = newPermissions
+        // diffResult.dispatchUpdatesTo(this)
+    }
+
+    private var locationPermissionEnabled: Boolean = false
+
+    private var locationPermissions: List<LocationPermissionEntity> = emptyList()
+        set(value) {
+            if (field != value) {
+                field = value
+                allowedLocationPermissions.clear()
+                deniedLocationPermissions.clear()
+                value.forEach {
+                    if (it.permission == LocationPermissionType.ALLOW_ONCE || it.permission == LocationPermissionType.ALLOW_ALWAYS) {
+                        allowedLocationPermissions.add(it)
+                    } else {
+                        deniedLocationPermissions.add(it)
+                    }
+                }
+            }
+        }
+
+    private var allowedLocationPermissions: MutableList<LocationPermissionEntity> = mutableListOf()
```

Comment:
It feels strange having these member variables declared throughout the file, and not all at the top

Reply:
it does feel strange, I'm moving them to the proper place